### PR TITLE
Fix filter images by repo

### LINKED ISSUE #1257: /images/json ignores repo name filter
Similar to #1244 but for repo name instead of labels

### PR DESCRIPTION
**Pull request by dnephin**
Fixes #1257

Support `/images/json`  filter by repo name/tag

### CODE REVIEW

File: test/integration/api/images.bats [hunk 1]
```diff
@@ -56,3 +56,15 @@ function teardown() {
 	[ "${#lines[@]}" -eq 2 ]
 	[[ "${lines[1]}" == *"image-with-labels"* ]]
 }
+
+@test "docker images imagetag" {
+	start_docker_with_busybox 2
+	swarm_manage
+
+	docker_swarm build -t testimage:latest $TESTDATA/imagelabel
```
Comment (dnephin) [inline test/integration/api/images.bats]: I tried using `docker tag` here but for some reason I'm seeing (what looks like) responses from the engine where `RepoTags` actually has tags from different repos (ex: `busybox:latest`, `testimage:latest`).

I haven't been able to reproduce this behaviour outside of the swarm test suite, so I'm not sure how that is happening.

This seems to work fine though.

**Comment by vieux** [comment]
@dnephin what about filtering at the API level so we won't have to touch the cluster interface ?

**Comment by dnephin** [comment]
That's certainly possible. What's wrong with touching the cluster interface?

The API layer kind of feels like the wrong place for this. Generally an api layer should only be for input validation, output formatting, handling http headers, and that kind of thing. I believe engine makes this distinction and tries to move the other logic into some other package. Moving it to the API also means that `parseImageName()` needs to become publish and the api needs to know more about image names.

It would be nice to have all the filtering happen in one place. Right now we have two of the filters in engine, and one in api. I think if the filtering by node that is done in api was actually done in engine, it would be a lot less code. It would be as simple as checking if it's own name was in the list of filtered nodes.  If not return.  Aggregation still needs to happen outside of engine, but maybe that could be part of the clustering code instead of api.

If the issue is having to deal with optional params (and golang's lack of keyword arguments). I saw a nice pattern in https://github.com/fsouza/go-dockerclient/blob/master/image.go#L85 of using a struct for the arg. That way if new ones get added the number of lines of code to change is still pretty minimal.


What do you think?

**Comment by vieux** [comment]
You're right, the API might not be the right place. 
When I look at the diff in `cluster/mesos/cluster.go` it feels wrong to put it in the interface.
I'd like to keep `Containers()` and `Images()` the same.

What about changing `Images(all bool, filters dockerfilters.Args) []*Image` to `Images() Images` introduce an Images type and put all filtering inside ?

**Comment by dnephin** [comment]
Cool, that sounds good

**Comment by vieux** [comment]
@dnephin  thanks.

**Comment by dnephin** [comment]
Ok, I made that change to move filtering to a new `Images` type.

I haven't moved the node filtering yet, but I can look at doing that next.

**Comment by dnephin** [comment]
I think this test failure is unrelated to my change. 

Let me know if this looks good.  I'd like to save the refactor of node filtering for another branch (since it's not touched by this change).

**Comment by dnephin** [comment]
@vieux how does this look?

File: api/handlers.go [hunk 2]
```diff
@@ -26,9 +26,10 @@ const APIVERSION = "1.16"
 
 // GET /info
 func getInfo(c *context, w http.ResponseWriter, r *http.Request) {
+	images := c.cluster.Images().Filter(cluster.ImageFilterOptions{})
```
Comment (vieux) [inline api/handlers.go]: you do need to declare this variable ?

Comment (dnephin) [inline api/handlers.go]: It's not necessary, but otherwise the line goes way over 80 char.

**Comment by vieux** [comment]
besides a small nit, LGTM thanks @dnephin

**Comment by dnephin** [comment]
Ok, I addressed the nit

@abronan it would be good to get this into 1.0.0 rc2 as well, for compose compatibility

**Comment by dnephin** [comment]
rebased on master

**Comment by abronan** [comment]
LGTM provided that the CI is green :)

**Comment by aluzzardi** [comment]
LGTM

**Comment by abronan** [comment]
Hmm there are some unrelated failures on `save`/`import`/`export`.

@vieux @jimmyxian not sure but looking at the failures, this seems like this is related to #1271? At least the tests are outdated with the image placement I guess.

**Comment by jimmyxian** [comment]
@abronan 
OK, I will check it..

**Comment by vieux** [comment]
LGTM

**Pull request merged**